packages/flutter_tools/lib/src/android/gradle_utils.dart

```diff
@@ -0,0 +1,277 @@
+import '../base/context.dart';
+import '../base/file_system.dart';
+import '../base/os.dart';
+import '../base/platform.dart';
+import '../base/terminal.dart';
+import '../base/utils.dart';
+import '../base/version.dart';
+import '../build_info.dart';
+import '../cache.dart';
+import '../globals.dart';
+import '../project.dart';
+import '../reporting/reporting.dart';
+import 'android_sdk.dart';
+import 'android_studio.dart';
+
+/// The environment variables needed to run Gradle.
+Map<String, String> get gradleEnv {
+  final Map<String, String> env = Map<String, String>.from(platform.environment);
+  if (javaPath != null) {
+    // Use java bundled with Android Studio.
+    env['JAVA_HOME'] = javaPath;
+  }
+  // Don't log analytics for downstream Flutter commands.
+  // e.g. `flutter build bundle`.
+  env['FLUTTER_SUPPRESS_ANALYTICS'] = 'true';
+  return env;
+}
+
+/// Gradle utils in the current [AppContext].
+GradleUtils get gradleUtils => context.get<GradleUtils>();
+
+/// Provides utilities to run a Gradle task,
+/// such as finding the Gradle executable or constructing a Gradle project.
+class GradleUtils {
+  /// Empty constructor.
+  GradleUtils();
+
+  /// Gets the Gradle executable path and prepares the Gradle project.
+  /// This is the `gradlew` or `gradlew.bat` script in the `android/` directory.
+  String getExecutable(FlutterProject project) {
+    final Directory androidDir = project.android.hostAppGradleRoot;
+    // Update the project if needed.
+    // TODO(egarciad): https://github.com/flutter/flutter/issues/40460.
+    migrateToR8(androidDir);
+    injectGradleWrapperIfNeeded(androidDir);
+
+    final File gradle = androidDir.childFile(
+      platform.isWindows ? 'gradlew.bat' : 'gradlew',
+    );
+    if (gradle.existsSync()) {
+      printTrace('Using gradle from ${gradle.absolute.path}.');
+      return gradle.absolute.path;
+    }
+    throwToolExit('Unable to locate gradlew script');
+    return null;
+  }
+}
+
+/// Migrates the Android's [directory] to R8.
+/// https://developer.android.com/studio/build/shrink-code
+@visibleForTesting
+void migrateToR8(Directory directory) {
+  final File gradleProperties = directory.childFile('gradle.properties');
+  if (!gradleProperties.existsSync()) {
+    throwToolExit('Expected file ${gradleProperties.path}.');
+  }
+  final String propertiesContent = gradleProperties.readAsStringSync();
+  if (propertiesContent.contains('android.enableR8')) {
+    printTrace('gradle.properties already sets `android.enableR8`');
+    return;
+  }
+  printTrace('set `android.enableR8=true` in gradle.properties');
+  try {
+    // Add `android.enableR8=true` to the next line in gradle.properties.
+    if (propertiesContent.isNotEmpty && !propertiesContent.endsWith('\n')) {
+      gradleProperties
+        .writeAsStringSync('\nandroid.enableR8=true\n', mode: FileMode.append);
+    } else {
+      gradleProperties
+        .writeAsStringSync('android.enableR8=true\n', mode: FileMode.append);
+    }
```

Comment:
this could be cleaner if the string wasn't repeated. Can we e.g. extract the string into a constant, or do something like
```dart
gradleProperties.writeAsStringSync(
  (propertiesContent.isNotEmpty && !propertiesContent.endsWith('\n')) ? '\n' : '')
  + 'android.enableR8=true\n',
  mode: FileMode.append,
);
￼
```

Reply:
I changed this a bit. Thanks for pointing it out :)